src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -609,13 +605,48 @@ internal TypeWithAnnotations InferCollectionElementType(BindingDiagnosticBag dia
             BoundExpression collectionExpr = this.GetBinder(collectionSyntax).BindValue(collectionSyntax, diagnostics, BindValueKind.RValue);
 
             var builder = new ForEachEnumeratorInfo.Builder();
-            GetEnumeratorInfoAndInferCollectionElementType(ref builder, ref collectionExpr, diagnostics, out TypeWithAnnotations inferredType);
+            GetEnumeratorInfoAndInferCollectionElementType(_syntax, collectionSyntax, ref builder, ref collectionExpr, IsAsync, diagnostics, out TypeWithAnnotations inferredType);
             return inferredType;
         }
+    }
+
+    partial class Binder
+    {
+        protected BoundExpression ConvertForEachCollection(
+            BoundExpression collectionExpr,
+            Conversion collectionConversionClassification,
+            TypeSymbol collectionType,
+            BindingDiagnosticBag diagnostics)
+        {
+            // We're wrapping the collection expression in a (non-synthesized) conversion so that its converted
+            // type (i.e. builder.CollectionType) will be available in the binding API.
+            Debug.Assert(!collectionConversionClassification.IsUserDefined);
+
+            BoundExpression convertedCollectionExpression = CreateConversion(collectionExpr.Syntax, collectionExpr, collectionConversionClassification, isCast: false, conversionGroupOpt: null, collectionType, diagnostics);
+
+            if ((convertedCollectionExpression as BoundConversion)?.Operand != (object)collectionExpr)
+            {
+                Debug.Assert(collectionConversionClassification.IsIdentity);
+                Debug.Assert(convertedCollectionExpression == (object)collectionExpr);
+
+                convertedCollectionExpression = new BoundConversion(
+                    collectionExpr.Syntax,
+                    collectionExpr,
+                    collectionConversionClassification,
+                    @checked: CheckOverflowAtRuntime,
+                    explicitCastInCode: false,
+                    conversionGroupOpt: null,
+                    ConstantValue.NotAvailable,
+                    collectionType);
+            }
+
+            return convertedCollectionExpression;
+        }
 
-        private bool GetEnumeratorInfoAndInferCollectionElementType(ref ForEachEnumeratorInfo.Builder builder, ref BoundExpression collectionExpr, BindingDiagnosticBag diagnostics, out TypeWithAnnotations inferredType)
+        // PROTOTYPE: Why is collectionExpr passed by ref?
+        protected bool GetEnumeratorInfoAndInferCollectionElementType(SyntaxNode syntax, ExpressionSyntax collectionSyntax, ref ForEachEnumeratorInfo.Builder builder, ref BoundExpression collectionExpr, bool isAsync, BindingDiagnosticBag diagnostics, out TypeWithAnnotations inferredType)
```

Comment:
Happy to see this is being done as part of collection literals! I don't remember how many times I went through this and ran away screaming.

Reply:
Could the collectionSyntax be of type SyntaxNode? Looks like we only fetch the location off of it.